Mathlib/Analysis/SpecialFunctions/Complex/PontryaginDuality.lean

```diff
@@ -0,0 +1,236 @@
+/-
+Copyright (c) 2023 Yaël Dillies, Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import Mathlib.Algebra.DirectSum.AddChar
+import Mathlib.Analysis.SpecialFunctions.Complex.CircleAddChar
+import Mathlib.GroupTheory.FiniteAbelian
+
+/-!
+# Pontryagin duality for finite abelian groups
+
+This file proves the Pontryagin duality in case of finite abelian groups. This states that any
+finite abelian group is canonically isomorphic to its double dual (the space of complex-valued
+characters of its space of complex-valued characters).
+
+We first prove it for `ZMod n` and then extend to all finite abelian groups using the
+Structure Theorem.
+-/
+
+noncomputable section
+
+open Circle Finset Function Multiplicative
+open Fintype (card)
+open scoped BigOperators DirectSum
+
+variable {α : Type*} [AddCommGroup α] {n : ℕ} {a b : α}
+
+namespace AddChar
+
+private def zmodAuxAux (n : ℕ) : ℤ →+ Additive Circle where
+  toFun x := .ofMul <| e <| x / n
+  map_zero' := by dsimp; rw [Int.cast_zero, zero_div, ofMul_eq_zero, map_zero_eq_one]
+  map_add' x y := by rw [← ofMul_mul, Equiv.apply_eq_iff_eq, Int.cast_add, add_div, map_add_eq_mul]
+
+@[simp]
+lemma zmodAuxAux_apply (n : ℕ) (z : ℤ) : zmodAuxAux n z = Additive.ofMul (e <| z / n) := rfl
+
+/-- The character sending `k : ZMod n` to `e ^ (2 * π * i * k / n)`. -/
+private def zmodAux (n : ℕ) : AddChar (ZMod n) Circle :=
+  AddChar.toAddMonoidHomEquiv.symm <| ZMod.lift n ⟨zmodAuxAux n, by
+    obtain hn | hn := eq_or_ne (n : ℝ) 0 <;> simp [hn, zmodAuxAux]⟩
+
+--TODO: Heavily generalise. Yaël's attempts at generalising failed :(
+@[simp] lemma aux (n : ℕ) (h) :
+    (⟨zmodAuxAux n, h⟩ : {f : ℤ →+ Additive Circle // f n = 0}) = zmodAuxAux n := rfl
+
+@[simp] lemma zmodAux_apply (n : ℕ) (z : ℤ) : zmodAux n z = e (z / n) := by simp [zmodAux]
+
+-- probably an evil lemma
+-- @[simp] lemma zmodAux_apply (n : ℕ) (x : ZMod n) : zmodAux n x = e (x / n) :=
+-- by simp [zmodAux]
+
+lemma zmodAux_injective (hn : n ≠ 0) : Injective (zmodAux n) := by
+  replace hn : (n : ℝ) ≠ 0 := Nat.cast_ne_zero.2 hn
+  simp [zmodAux, ZMod.lift_injective, CharP.intCast_eq_zero_iff _ n, e_eq_one, div_eq_iff hn,
+    mul_comm _ (n : ℝ), -forall_exists_index]
+  norm_cast
+  exact fun _ ↦ id
+
+/-- Indexing of the complex characters of `ZMod n`. `AddChar.zmod n x` is the character sending `y`
+to `e ^ (2 * π * i * x * y / n)`. -/
+def zmod (n : ℕ) (x : ZMod n) : AddChar (ZMod n) Circle :=
+  (zmodAux n).compAddMonoidHom <| AddMonoidHom.mulLeft x
```

Comment:
Is it any better to write this in terms of `toCircle`? It's a group hom in ZMod n composed with that; maybe this also saves some of the lemmas above

Reply:
It is better!